File: pd_combo.py
# -*- coding: utf-8 -*-
# vim:set foldmethod=marker:

import random
import pprint
import copy
import pprint
import operator
import time
import numpy as np
import sys
import numba

ROW = 5
COL = 6
#MAX_TURN = 40
MAX_TURN = 30
#BEAM_WIDTH = 5000
BEAM_WIDTH = 100

class member(): # route pattern{{{
    def __init__(self, nowC, nowR, prev=-1, length=MAX_TURN):
        self.score = 0  # combo number
        self.nowC  = nowC # now column
        self.nowR  = nowR # now row
        self.prev = prev # previous position (1:up, 2:down, 3:left, 4:right, -1:None)
        self.movei = [[-1] * 2] * length
        #self.movei = np.full(length * 2, -1, dtype=np.int8).reshape(length, 2)
        # }}}

class pd_combo():
    def __init__(self, row=5, col=6, field=None, max_turn=MAX_TURN, beam_width=BEAM_WIDTH):# {{{
        self.row = row
        self.col = col
        self.max_turn = max_turn
        self.beam_width = BEAM_WIDTH
        if field != None:
            self.field = field
        else:
            #self.field = [[0 for _ in range(self.col)] for _ in range(self.row)]
            #self.field = [[0] * self.col] * self.row
            self.field = np.zeros(self.row * self.col, dtype=np.int8).reshape(self.row, self.col)
        #self.chainflag = [[0 for _ in range(self.col)] for _ in range(self.row)]
        #self.chainflag = [[0] * self.col] * self.row
        self.chainflag = np.zeros(self.row * self.col, dtype=np.int8).reshape(self.row, self.col)
        #self.dummy = [[0 for _ in range(self.col)] for _ in range(self.row)]
        #self.dummy = [[0] * self.col] * self.row
        self.dummy = np.zeros(self.row * self.col, dtype=np.int8).reshape(self.row, self.col)
        #self.t_erace = [[0 for _ in range(self.col)] for _ in range(self.row)]
        #self.t_erace = [[0] * self.col] * self.row
        self.t_erace = np.zeros(self.row * self.col, dtype=np.int8).reshape(self.row, self.col)
        self.max_count = 0  # chained drop count
        self.route = [[-1] * 2] * self.max_turn
        #self.f_field = [[0 for _ in range(self.col)] for _ in range(self.row)]
        #self.f_field = [[0] * self.col] * self.row
        self.f_field = np.zeros(self.row * self.col, dtype=np.int8).reshape(self.row, self.col)
        self.max_count = 0  # chained drop count
        self.d_1 = 0 # fire
        self.d_2 = 0 # water
        self.d_3 = 0 # wood
        self.d_4 = 0 # light
        self.d_5 = 0 # dark
        self.d_6 = 0 # cure
# }}}

    def make_field(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                self.field[i][j] = random.randint(0, 6)# }}}

    def show_field(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.field[i, j], end="")
            print("\n", end="")# }}}

    def show_field2(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.field[i][j], end="")
            print("\n", end="")# }}}

    def show_f_field(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.f_field[i][j], end="")
            print("\n", end="")# }}}

    def show_chainflag(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.chainflag[i][j], end="")
            print("\n", end="")# }}}

    def show_dummy(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.dummy[i][j], end="")
            print("\n", end="")# }}}

    def show_t_erace(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                print(self.t_erace[i][j], end="")
            print("\n", end="")# }}}

    def fall(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                for k in range(self.row):
                    if k+1 == self.row:
                        break
                    if self.field[k+1][j] == 0:
                        self.field[k+1][j] = self.field[k][j]
                        self.field[k][j] = 0# }}}

    def set(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                if self.field[i][j] == 0:
                    self.field[i][j] = random.randint(1, 6)
        # }}}

    def count_drop(self):# {{{
        for i in range(self.row):
            for j in range(self.col):
                if self.field[i][j] == 1:
                    self.d_1 += 1
                elif self.field[i][j] == 2:
                     self.d_2 += 1
                elif self.field[i][j] == 3:
                     self.d_3 += 1
                elif self.field[i][j] == 4:
                     self.d_4 += 1
                elif self.field[i][j] == 5:
                     self.d_5 += 1
                elif self.field[i][j] == 6:
                     self.d_6 += 1
         # }}}

    def swap(self, i1, j1, i2, j2):# {{{
        temp = self.field[i1][j1]
        self.field[i1][j1] = self.field[i2][j2]
        self.field[i2][j2] = temp# }}}

    def operation(self):# {{{
        now_col = self.route[0][0]
        now_row = self.route[0][1]

        for i in range(self.max_turn):
            if self.route[i][0] == -1 or self.route[i][1] == -1:
                break
            self.swap(now_row, now_col, self.route[i][1], self.route[i][0])
            #print("movei[i]:"+str(i))
            #self.show_field()
            now_col = self.route[i][0]
            now_row = self.route[i][1]

        # }}}

    def chain(self, now_row, now_col, drop, count):# {{{
        # if out of filed range, return
        if now_row == -1 or now_row == self.row or now_col == -1 or now_col == self.col:
            return
        # if same color drop is not searced
        if self.field[now_row][now_col] == drop and self.chainflag[now_row][now_col] == 0:
            self.chainflag[now_row][now_col] = -1   # searched
            if self.max_count < count:
                self.max_count = count
            self.dummy[now_row][now_col] = -1

            self.chain(now_row - 1, now_col, drop, count + 1)
            self.chain(now_row + 1, now_col, drop, count + 1)
            self.chain(now_row, now_col - 1, drop, count + 1)
            self.chain(now_row, now_col + 1, drop, count + 1)# }}}

    def evaluate(self):# {{{
        value = 0
        self.chainflag = [[0 for _ in range(self.col)] for _ in range(self.row)]
        for i in range(self.row):
            for j in range(self.col):
                if self.chainflag[i][j] == 0 and self.field[i][j] != 0:
                    max_count = 0
                    self.dummy = [[0 for _ in range(self.col)] for _ in range(self.row)]
                    self.chain(i, j, self.field[i][j], 1)
                    if self.max_count >= 3:
                        if self.check() == 1:
                            value += 1
        return value# }}}

    def check(self):# {{{
        # if drop chain 3 times over, return 1
        v = 0
        for i in range(self.row):
            for j in range(self.col - 2):
                if self.dummy[i][j] == -1 and self.dummy[i][j + 1] == -1 and self.dummy[i][j + 2] == -1 and \
                 self.field[i][j] == self.field[i][j + 1] and self.field[i][j] == self.field[i][j +2]:
                    self.t_erace[i][j] = -1
                    self.t_erace[i][j + 1] = -1
                    self.t_erace[i][j + 2] = -1
                    v = 1

        for i in range(self.row - 2):
            for j in range(self.col):
                if self.dummy[i][j] == -1 and self.dummy[i + 1][j] == -1 and self.dummy[i + 2][j] == -1 and \
                 self.field[i][j] == self.field[i + 1][j] and self.field[i][j] == self.field[i + 2][j]:
                    self.t_erace[i][j] = -1
                    self.t_erace[i + 1][j] = -1
                    self.t_erace[i + 2][j] = -1
                    v = 1
        return v  # 0: 0 combo, 1: some combo
    # }}}

    def sum_e(self): # {{{
        # consider "otoshi",  not consider "ochikon" , and caliculate combo
        combo = 0
        while(1):
            self.t_erace = [[0 for _ in range(self.col)] for _ in range(self.row)]
            a = self.evaluate()
            if a == 0:  # when 0 combo, break
                break
            for i in range(self.row):
                for j in range(self.col):
                    if self.t_erace[i][j] == -1:
                        self.field[i][j] = 0  # clear combo drop
            self.fall()
            combo += a
        return combo # }}}

    def beam_search(self):# {{{
        queue = []
        for i in range(self.row):
            for j in range(self.col):
                cand = member(j, i, -1)
                cand.movei[0]= [j, i]
                queue.append(cand)

        dx = [-1,  0, 0, 1]
        dy = [ 0, -1, 1, 0]

        for p in range(1,self.max_turn):
            pqueue = []
            for q in range(len(queue)):
                temp = copy.deepcopy(queue[q])
                for j in range(len(dx)):
                    self.field = copy.deepcopy(self.f_field)
                    cand = copy.deepcopy(temp)
                    if ( 0 <= temp.nowC + dx[j] < self.col and \
                            0 <= temp.nowR + dy[j] < self.row ):
                        if cand.prev + j == 3:
                            continue
                        cand.nowC = temp.nowC + dx[j]
                        cand.nowR = temp.nowR + dy[j]
                        cand.movei[p] = [cand.nowC, cand.nowR]
                        #print("q:"+str(q)+", p:"+str(p)+", cand.nowC:"+str(cand.nowC)+", cand.nowR:"+str(cand.nowR))
                        #print("cand.movei", end="")
                        #print(cand.movei)
                        self.route = copy.deepcopy(cand.movei)
                        self.operation()
                        #print("operated field")
                        #self.show_field()
                        cand.score = self.sum_e()
                        #print("cand.score:"+str(cand.score))
                        cand.prev = j
                        pqueue.append(cand)

            pqueue.sort(key=operator.attrgetter("score"), reverse=True)
            #print("pqueue[0].score:"+str(pqueue[0].score))
            #print("pqueue[1].score:"+str(pqueue[1].score))

            queue = []
            if self.beam_width < len(pqueue):
                width = self.beam_width
            else:
                width = len(pqueue)
            for i in range(width):
                queue.append(pqueue[i])
            #print("len(queue):"+str(len(queue)))

        return queue[0] # best route(=movei)

# }}}

if __name__ == "__main__":

    avg = 0
    avg_time = time.time() - time.time()
    repeat_count = 1
    for i in range(repeat_count):
        start_time = time.time()
        #pd_cmb = pd_combo(3, 4)
        #pd_cmb = pd_combo(4, 5)
        pd_cmb = pd_combo(ROW, COL)
        #print("pd_cmb.col:"+str(pd_cmb.col))
        #print("pd_cmb.row:"+str(pd_cmb.row))
        #print(pd_cmb.field)
        #print("pd_cmb.field size:"+str(sys.getsizeof(pd_cmb.field)))
        #print(pd_cmb.field.dtype)
        #print("\ninitial field")
        #pd_cmb.show_field()
        pd_cmb.set()  # make initial filed
        print("\nsetted field")
        pd_cmb.show_field()
        #pd_cmb.show_field2()

        pd_cmb.f_field = copy.deepcopy(pd_cmb.field)  # copy initial field
        best_member = pd_cmb.beam_search()  # best member(route, score,,,)
        #print("(x, y): ("+str(best_member.movei[0][0])+", "+str(best_member.movei[0][1])+")")
        #print(best_member.movei)

        #for i in range(pd_cmb.max_turn):
        #    if best_member.movei[i][0] == -1 or best_member.movei[i][1] == -1 :
        #        break
        #    if best_member.movei[i][0] == (best_member.movei[i - 1][0] + 1):
        #        print("RIGHT")
        #    if best_member.movei[i][0] == (best_member.movei[i - 1][0] - 1):
        #        print("LEFT")
        #    if best_member.movei[i][1] == (best_member.movei[i - 1][1] + 1):
        #        print("UP")
        #    if best_member.movei[i][1] == (best_member.movei[i - 1][1] - 1):
        #        print("DOWN")
        #    print("\n")
        pd_cmb.field  = copy.deepcopy(pd_cmb.f_field)
        pd_cmb.route = copy.deepcopy(best_member.movei)
        pd_cmb.operation()
        print("\noperated field")
        pd_cmb.show_field()
        elapsed_time = time.time() - start_time
        #print("elapsed_time:{0}".format(elapsed_time) + "[sec]")
        avg += best_member.score
        avg_time += elapsed_time
    print("avarage score:"+str(avg/repeat_count))
    print("avarage time:{0}".format(avg_time/repeat_count) + str("[sec]"))

File: pd_combo.jl
# -*- coding: utf-8 -*-
# vim:set foldmethod=marker:

ROW = 5
COL = 6
MAX_TURN = 70
BEAM_WIDTH = 5000

field = zeros(Int8, ROW, COL)
f_field = zeros(Int8, ROW, COL)
chainflag = zeros(Int8, ROW, COL)
dummy = zeros(Int8, ROW, COL)
t_erace = zeros(Int8, ROW, COL)
max_count = 0
route = fill(-1, MAX_TURN, 2)

mutable struct member#={{{=#
    score::Int8
    nowR::Int8
    nowC::Int8
    prev::Int8
    movei
end#=}}}=#

function sort_member(a, b)#={{{=#
    if a.score < b.score
        return true
    end
    return false
end#=}}}=#

# sort example{{{{{{
#    m1 = member(1, 1, 3, 4, [1 1; 2 2;])
#    m2 = member(3, 2, 3, 4, [])
#    m3 = member(2, 3, 3, 4, [])
#    arr_member = Array{member}(undef, 3)
#    arr_member[1] = m1
#    arr_member[2] = m2
#    arr_member[3] = m3
#    sort!(arr_member, lt=sort_member, rev=true)
#}}}}}}

function set(field1=[])#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    if field1 == []
	    for i in 1:ROW
	    	for j in 1:COL
	    		if field[i, j] == 0
	    			field[i, j] = rand(1:6)
	    		end
	    	end
	    end
    else
	    for i in 1:ROW
	    	for j in 1:COL
                w = (i-1)*COL+j
                field[i, j] = parse(Int8, field1[(i-1)*COL+j])
	    	end
	    end
    end
end#=}}}=#

function show(field)#={{{=#
    for i in 1:ROW
        for j in 1:COL
            print(field[i, j])
        end
        print("\n")
    end
end#=}}}=#

function fall()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    for i in 1:ROW
        for j in 1:COL
            for k in 1:ROW
                if k+1 > ROW
                    break
                end
                if field[k+1, j] == 0
                    global field[k+1, j] = field[k, j]
                    global field[k, j] = 0
                end
            end
        end
    end
end#=}}}=#

function swap(i1, j1, i2, j2)#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    temp = field[i1, j1]
    global field[i1, j1] = field[i2, j2]
    global field[i2, j2] = temp
end#=}}}=#

function operation()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    now_col = route[1, 2]
    now_row = route[1, 1]
    #print("now_col" , now_col , ", now_row:" , now_row, "\n")

    for i in 1:MAX_TURN
        if route[i, 1] == -1 || route[i, 2] == -1
            break
        end
        swap(now_col, now_row, route[i, 2], route[i, 1])
        now_col = route[i, 2]  # x position
        now_row = route[i, 1]  # y position
    end
end#=}}}=#

function chain(now_row, now_col, drop, count)#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    if now_row == 0 || now_row > ROW || now_col == 0 || now_col > COL
        return
    end
    if field[now_row, now_col] == drop && chainflag[now_row, now_col] == 0
        global chainflag[now_row, now_col] = -1
        if max_count < count
            global max_count = count
        end
        global dummy[now_row, now_col] = -1

        chain(now_row-1, now_col, drop, count+1)
        chain(now_row+1, now_col, drop, count+1)
        chain(now_row, now_col-1, drop, count+1)
        chain(now_row, now_col+1, drop, count+1)
    end
end#=}}}=#

function evaluate()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    value = 0
    chainflag = zeros(Int8, ROW, COL)
    for i in 1:ROW
        flg_row = 0
        for j in 1:COL
            if chainflag[i, j] == 0 && field[i, j] != 0
                global max_count = 0
                global dummy = zeros(Int8, ROW, COL)
                chain(i, j, field[i, j], 1)
                if max_count >= 3
                    if check() == 1
                        value += 1
                    end
                end
            end
            if j <= COL-1
                if field[i, j] != 0 && field[i, j] == field[i, j+1]
                    flg_row += 1
                    #println("flg_row:", flg_row)
                end
            end
        end
        if flg_row == COL-1
            value += 10
        end
    end
    return value
end#=}}}=#

function check()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    v = 0
    for i in 1:ROW
        for j in 1:COL-2
            if dummy[i, j] == -1 && dummy[i, j+1] == -1 && dummy[i, j+2] == -1 && field[i, j] == field[i, j+1] && field[i, j] == field[i, j+2]
                global t_erace[i, j] = -1
                global t_erace[i, j+1] = -1
                global t_erace[i, j+2] = -1
                v = 1
            end
        end
    end

    for i in 1:ROW-2
        for j in 1:COL
            if dummy[i, j] == -1 && dummy[i+1, j] == -1 && dummy[i+2, j] == -1 && field[i, j] == field[i+1, j] && field[i, j] == field[i+2, j]
                global t_erace[i, j] = -1
                global t_erace[i+1, j] = -1
                global t_erace[i+2, j] = -1
                v = 1
            end
        end
    end
    return v
end#=}}}=#

function sum_e()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
    combo = 0
    while(1==1)
        global t_erace = zeros(Int8, ROW, COL)
        a = evaluate()
        if a == 0
            break
        end
        for i in 1:ROW
            for j in 1:COL
                if t_erace[i, j] == -1
                    global field[i, j] = 0
                end
            end
        end
        fall()
        combo += a
    end
    return combo
end#=}}}=#

function beam_search()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route

    que_member = member[]
    temp_que = member[]
    i = 1
    for i in 1:ROW
        for j in 1:COL
            cand = member(0, i, j, -1, [])
            cand.movei = fill(-1, MAX_TURN, 2)
            cand.movei[1, :] = [i j]
            push!(que_member, cand)
        end
    end
    #println("size(que_member, 1):", size(que_member, 1))
    #println("que_member[1]:", que_member[1])

    dx = [-1;  0; 0; 1]
    dy = [ 0; -1; 1; 0]
    max_value = 0
    width = 0

    for i in 1:MAX_TURN
        #pque_member = Array{member}(undef, 1)
        pque_member = member[]
        while length(que_member) != 0
            #temp = pop!(que_member)
            temp = que_member[1]
            #println("before length(que_member)", length(que_member))
            temp_que = copy(que_member[2:end])
            que_member = copy(temp_que)
            #println("after length(que_member)", length(que_member))
            #println("temp:", temp)
            for j in 1:length(dx)
                field = copy(f_field)
                cand = member(0, i, j, -1, [])
                cand.score = temp.score
                cand.nowC = temp.nowC
                cand.nowR = temp.nowR
                cand.prev = temp.prev
                cand.movei = copy(temp.movei)
                if ( 1 <= temp.nowC + dx[j] <= COL && 1 <= temp.nowR + dy[j] <= ROW )
                    if cand.prev + j == 5
                        continue
                    end
                    cand.nowC = temp.nowC + dx[j]
                    cand.nowR = temp.nowR + dy[j]
                    #cand.movei[i, :] = [cand.nowC cand.nowR;]
                    cand.movei[i, :] = [temp.nowC+dx[j] temp.nowR+dy[j]]
                    #println("temp.nowC:", temp.nowC, ", temp.nowR:", temp.nowR, ", dx[j]:", dx[j], ", dy[j]:", dy[j])
                    #println("temp.nowC + dx[j]:", temp.nowC + dx[j])
                    #println("temp.nowR + dy[j]:", temp.nowR + dy[j])
                    #println("cand.movei[i, :] in beam_search:", cand.movei[i, :])
                    route = copy(cand.movei)
                    #println("route in beam_search:", route)
                    operation()
                    cand.score = sum_e()
                    #println("cand.score in beam_search:", cand.score)
                    cand.prev = j
                    push!(pque_member, cand)
                end
            end
        end
        sort!(pque_member, lt=sort_member, rev=true)

        if MAX_TURN < length(pque_member)
            width = MAX_TURN
        else
            width = length(pque_member)
        end
        que_member = Array{member}(undef, width)
        que_member[1:width, : ] = pque_member[1:width, : ]
    end

    #println(que_member[1:10, : ])
    return que_member[1]  # return best_member

    best_member = que_member[1]
    return best_member
end#=}}}=#

function main()#={{{=#
    global field, f_field, chainflag, dummy, t_erace, max_count, route
	set()
    #set("315211554451322114424566531621")  # 15 combo
    println("initial field.")
    global field, f_field, route
    show(field)
    f_field = copy(field)
    #route[1:6 , : ] = [1 1; 2 1; 3 1; 4 1; 5 1; 6 1;]
    #print(route[1:6, :], "\n")

    #best_member = member()
    best_member = beam_search()
    println("best_member:",  best_member)
    #println(size(best_member.movei, 1))
    route[1:size(best_member.movei, 1), : ] = best_member.movei
    #println("route:",  route)
    field = copy(f_field)
    operation()
    println("after operation")
    show(field)
    combo = sum_e()
    println("combo:", combo)
    println("after sum_e")
    show(field)

end#=}}}=#

main()

